### ISSUE TITLE
Produce an "insert header" quick fix for undeclared functions like `printf` in C code

### ISSUE THREAD

**Issue by reportaman** [clangd/clangd#937]
Is there a key binding that can be used to search the header files and suggest to include the missing ones?

CLion has a feature that does that, check the video at time 00:40 https://youtu.be/Srnw1dI1iAA?t=40

**Comment by kirillbobyrev** [comment]
That is indeed a very useful feature! I have been working on solving this problem for several months now. We have decided to split the feature into two: unused headers and missing headers. Right now, I'm actively developing the "unused headers" diagnostic and, if everything goes well, will enable it by default in the next major LLVM release. Realistically, "missing includes" is not started yet and is probably few months away. We still need to solve some important problems in this space (the semantics of "used header" are very different across the projects and guidelines, and it's also unfortunate if the diagnostics break the build for some reason; another problem is that visibility is a very tricky concept in C++).

Sadly, there are no docs about this right now and I should certainly write some. If you're curious, the feature is called IncludeCleaner and the core of it is in [IncludeCleaner.cpp](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/IncludeCleaner.cpp). We're currently experimenting with the feature and using it ourselves; you can do that too! In order to use the feature, build/download latest clangd binary (e.g. from our [nightly releases](https://github.com/clangd/clangd/releases)) and put the following into your clangd configuration file (more on clangd config [here](https://clangd.llvm.org/config.html)):

```yaml
Diagnostics:
  UnusedIncludes: Strict
```

Please note that this does not have 100% accuracy right now but we try to be conservative and avoid offering wrong diagnostics (leaning towards conservatively _not_ marking headers that might be unsused rather than being more agressive and marking more candidates as unused which can actually break the build). If you encounter any issues with IncludeCleaner, please open the bugs under [clangd/clangd#issues](https://github.com/clangd/clangd) and assign them to me!

**Comment by HighCommander4** [comment]
> CLion has a feature that does that, check the video at time 00:40 https://youtu.be/Srnw1dI1iAA?t=40

What I'm seeing in the video is:

  * the user has a file which does not yet include `<string>`
  * the user adds a declaration `std::string name;` to the file
  * the `std::string` gets a diagnostic
  * with the cursor over the diagnostic, you can invoke an editor action that inserts the `<string>` include for you

As far as I can tell, clangd supports basically the same thing today: if you hover over the diagnostic, there is a **quick fix** suggested which inserts the `<string>` include. You can also activate it using a keyboard shortcut (this part is editor-specific).

**Comment by HighCommander4** [comment]
In fact, I would argue the workflow for the above scenario is **even better** in clangd: if, at the time you're typing `std::string`, you select it from the completion pop-up rather than fully typing it out, then the include is automatically added as part of accepting the completion proposal.

**Comment by reportaman** [comment]
@HighCommander4 Oh I know this could be off topic to ask for specific editor, and plugin. Though do you know how that works with coc-clangd in neovim? What keybinding would I need to trigger the action of automatically fixing a problem like missing header file?

I tried a bunch of "code actions" mappings from my websearch, none of it seems to work. The message I get is "code action not available". Am not sure what I could be doing wrong...

```vim
"nmap <leader>qf <Plug>(coc-fix-current)
"nmap <leader>qf <Plug>(coc-codeaction-cursor)
nmap <leader>qf v<Plug>(coc-codeaction-selected)
```
I also look forwarded to the more advanced features suggested by @kirillbobyrev!

**Comment by HighCommander4** [comment]
> @HighCommander4 Oh I know this could be off topic to ask for specific editor, and plugin (though I had posted this question originally in coc-clangd). Do you know how that works with coc-clangd in neovim? What keybinding would I need to trigger the action of automatically fixing a problem like missing header file?

I don't use vim, but looking at coc.nvim issues like https://github.com/neoclide/coc.nvim/issues/2920, maybe `coc-fix-current`?

**Comment by reportaman** [comment]
@HighCommander4 It doesn't work on my end, it keeps on giving the same error: "no code action available"... @fannheyward, any idea??

**Comment by HighCommander4** [comment]
@reportaman To rule out an issue on the clangd side, can you share a clangd log taken with `--log=verbose`? I'd like to confirm that clangd is in fact sending the quick fix over LSP.

**Comment by reportaman** [comment]
@HighCommander4 Before that, my coc-settings.json file does not have any clangd specific settings. Am on a Apple Silicon ARM64 Darwin machine & the standard header files are not in the default location (but it LSP seems to be able to suggest the correct headers for autocompletion). I guess, you want me to edit and add ``clangd.arguments`` option mentioned on page https://github.com/clangd/coc-clangd? Where would it generate the logs??

**Comment by HighCommander4** [comment]
> I guess, you want me to edit and add `clangd.arguments` option mentioned on page https://github.com/clangd/coc-clangd?

Yep, add `--log-=verbose` to that.

> Where would it generate the logs??

clangd sends them to stderr. Not sure where coc.nvim surfaces the language server's stderr...

**Comment by HighCommander4** [comment]
> Not sure where coc.nvim surfaces the language server's stderr...

Maybe using https://github.com/neoclide/coc.nvim/wiki/F.A.Q#how-to-get-log-of-cocnvim (or maybe that log shows the LSP traffic logged independently by coc.nvim, that also works)

**Comment by reportaman** [comment]
@HighCommander4 I did just that, and this is the log I got. It looks like coc's log not clangds.

```
2021-11-26T17:23:51.180 DEBUG (pid:19330) [workspace] - buffer created 1
2021-11-26T17:23:51.182 DEBUG (pid:19330) [workspace] - buffer created 2
2021-11-26T17:23:51.200 DEBUG (pid:19330) [sources] - created service source snippets
2021-11-26T17:23:51.201 DEBUG (pid:19330) [sources] - created service source snippets-source
2021-11-26T17:23:51.208 INFO (pid:19330) [coc-tabnine] - Using tabnine from /Users/amanmehra/.config/coc/extensions/coc-tabnine-data/binaries/4.0.9/aarch64-apple-darwin/TabNine
2021-11-26T17:23:51.208 DEBUG (pid:19330) [sources] - created service source tabnine
2021-11-26T17:23:51.210 INFO (pid:19330) [plugin] - coc.nvim 0.0.80-19a630de32 initialized with node: v17.0.1 after 56ms
2021-11-26T17:23:55.952 DEBUG (pid:19330) [workspace] - buffer unload 2
2021-11-26T17:23:56.032 DEBUG (pid:19330) [workspace] - buffer created 3
2021-11-26T17:23:56.034 INFO (pid:19330) [services] - registered service "clangd"
2021-11-26T17:23:56.034 DEBUG (pid:19330) [services] - starting service: clangd
2021-11-26T17:23:56.035 INFO (pid:19330) [services] - clangd state change: stopped => starting
2021-11-26T17:23:56.039 INFO (pid:19330) [language-client-index] - Language server "clangd" started with 19338
2021-11-26T17:23:56.062 INFO (pid:19330) [services] - clangd state change: starting => running
2021-11-26T17:23:56.063 DEBUG (pid:19330) [sources] - created service source clangd
2021-11-26T17:23:56.066 INFO (pid:19330) [services] - service clangd started
2021-11-26T17:24:04.057 INFO (pid:19330) [attach] - receive notification: doQuickfix []
2021-11-26T17:24:04.064 ERROR (pid:19330) [attach] - Notification error: doQuickfix [] Error: No quickfix action available
    at CodeActions.doQuickfix (/Users/amanmehra/.vim/plugged/coc.nvim/build/index.js:63485:17)
    at async Plugin.cocAction (/Users/amanmehra/.vim/plugged/coc.nvim/build/index.js:68303:19)
    at async NeovimClient2.<anonymous> (/Users/amanmehra/.vim/plugged/coc.nvim/build/index.js:68413:15)
2021-11-26T17:24:30.139 INFO (pid:19330) [attach] - receive notification: openLog []

```

**Comment by reportaman** [comment]
Also, if I add another argument like ``2>~/clangd.log``, the server fails to start...

**Comment by HighCommander4** [comment]
It looks like coc's logs do not contain the LSP traffic. Let's try this one: https://github.com/neoclide/coc.nvim/wiki/Debug-language-server#using-output-channel

**Comment by reportaman** [comment]
@HighCommander4  Perhaps this??

```
V[17:38:05.270] <<< {"id":2,"jsonrpc":"2.0","method":"textDocument/codeAction","params":{"context":{"diagnostics":[{"code":"-Wimplicit-function-declaration","message":"Implicitly declaring library function 'printf' with type 'int (const char *, ...)'","range":{"end":{"character":10,"line":9},"start":{"character":4,"line":9}},"relatedInformation":[{"location":{"range":{"end":{"character":10,"line":9},"start":{"character":4,"line":9}},"uri":"file:///Users/reportaman/dev/learning/c/pointers/1_intptr.c"},"message":"Include the header <stdio.h> or explicitly provide a declaration for 'printf'"}],"severity":2,"source":"clang"}],"only":["quickfix"]},"range":{"end":{"character":0,"line":10},"start":{"character":0,"line":9}},"textDocument":{"uri":"file:///Users/reportaman/dev/learning/c/pointers/1_intptr.c"}}}

I[17:38:05.271] <-- textDocument/codeAction(2)
V[17:38:05.271] ASTWorker running EnumerateTweaks on version 1 of /Users/reportaman/dev/learning/c/pointers/1_intptr.c
I[17:38:05.271] --> reply:textDocument/codeAction(2) 0 ms
V[17:38:05.271] >>> {"id":2,"jsonrpc":"2.0","result":[]}

[Trace - 5:38:05 PM] Received response 'textDocument/codeAction - (2)' in 3ms.
Result: []


I[17:38:05.271] --> textDocument/clangd.fileStatus
V[17:38:05.271] >>> {"jsonrpc":"2.0","method":"textDocument/clangd.fileStatus","params":{"state":"idle","uri":"file:///Users/reportaman/dev/learning/c/pointers/1_intptr.c"}}

[Trace - 5:38:05 PM] Received notification 'textDocument/clangd.fileStatus'.
Params: {
    "state": "idle",
    "uri": "file:///Users/reportaman/dev/learning/c/pointers/1_intptr.c"
}

```

**Comment by HighCommander4** [comment]
Yep, that's the kind of output I'm looking for. But please provide the entire log (I'm looking specifically for the `textDocument/publishDiagnostics` message where clangd sends the diagnostic associated with the unresolved name), and also the exact source code to cross-reference

**Comment by reportaman** [comment]
@HighCommander4 Here we go: [clangd.log](https://github.com/clangd/clangd/files/7611026/clangd.log)

**Comment by HighCommander4** [comment]
Thanks. So, for clarity, the code is:

```c
#include <stddef.h>
#include <stdlib.h>
int main() {
  int a;
  int *pa;
  int **ppa;
  a = 20;
  pa = &a;
  ppa = &pa;
  printf("int a = %d, int *pa=%p, int **ppa=%p\n", a, pa, ppa);
  return EXIT_SUCCESS;
}
```

and this is a C (not C++) source file.

In C, using a function like `printf` without declaring it is not ill-formed. clang does produce a diagnostic, but it's a warning, not an error, and it indeed does not come with a quick fix.

Perhaps clang/clangd could do better on this example, and provide a quick fix even in the case of a warning like this.

**Comment by HighCommander4** [comment]
**Summary:** The issue is on the clangd side, but it's specific to C code, and to functions (an unresolved **type** like `time_t` **does** get a quick fix even in C mode).

**Comment by HighCommander4** [comment]
Possible duplicate of #355

**Comment by reportaman** [comment]
@HighCommander4 Yep, I am working with C code exclusively. If I move the file to .cpp, it does show it as an error and code- action does fix it by adding a ``#include <cstdio>`` among headers (in case of C, it would need to add``#include <stdio.h>`` I guess, at least that's what I am used to in C).

**Comment by reportaman** [comment]
@HighCommander4 There seems to be a difference in what you mention is already implemented, and what this ticket asked (and is showcased in CLion video) as per the comment https://github.com/clangd/coc-clangd/issues/376#issuecomment-980800662. So I guess @kirillbobyrev might be correct that it's not implemented yet?

**Comment by HighCommander4** [comment]
> @HighCommander4 There seems to be a difference in what you mention is already implemented, and what this ticket asked (and is showcased in CLion video) as per (1) in comment [clangd/coc-clangd#376 (comment)](https://github.com/clangd/coc-clangd/issues/376#issuecomment-980800662).

It's true that the presence of header insertion quick fix is contingent on clangd already having the missing symbol in its index, which generally requires that some other file in the project include it already.

For standard library symbols, there is a plan to proactively index them, tracked in https://github.com/clangd/clangd/issues/403.

> So I guess @kirillbobyrev might be correct that it's not implemented yet?

The "missing headers" feature that @kirillbobyrev described sounds to me like a related but different feature, which:

 1. can be invoked anywhere in the file (cursor does not have to be over a missing symbol)
 2. inserts **all** missing headers
 3. possibly also tweaks the set of included headers to obey include-what-you-use rules (i.e. if a header was included indirectly but you use a symbol from it directly, it would insert the header directly, even though this is not required to resolve any diagnostic)

**Comment by kirillbobyrev** [comment]
Right, sorry, I wasn't very clear and apparently I didn't understand the CLion behavior correctly :( @HighCommander4 thanks for correcting me!

**Comment by reportaman** [comment]
@sam-mccall Thanks! it now does add missing header for printf, but it in the process overwrites the function prototype, and duplicates the printf statement.

Also, there is no fix for EXIT_SUCCESS, but I guess that's because the current attempt focusses on functions rather than all symbols from standard headers.

Here's the video:

https://user-images.githubusercontent.com/71736629/145337253-5f77e119-3f6a-48e8-b7cb-7a38f16edb14.mov

**Comment by reportaman** [comment]
Actually, uninstalling and reinstalling coc-clangd removes the problem! It works well now, thanks so much :)

**Comment by HighCommander4** [comment]
`EXIT_SUCCESS` does work me, keeping mind the caveat discussed in [this comment](https://github.com/clangd/clangd/issues/937#issuecomment-980973009) that **another** file in the project needs to include the relevant header (in this case, `<stdlib.h>`) so that clangd knows about it.

**Comment by reportaman** [comment]
@HighCommander4 For me stdio, and sodlib.h worked with the functions. For sodlib.h worked for malloc(), but not for EXIT_SUCCESS. Also, nothing from time.h seems to work, time() etc functions.

I am not sure what project would mean in my case, I just have a folder of independent c files that are independently compiled.

**Comment by HighCommander4** [comment]
> I am not sure what project would mean in my case, I just have a folder of independent c files that are independently compiled.

Clangd has a background indexer feature, where it goes through your `compile_commands.json` if you have one and parses every file and maintains a symbol index that allows e.g. find-references to work across translation units.

My understanding is that the ability to offer an include-insertion quick fix for a symbol depends on the symbol being present in the index already.

**Comment by reportaman** [comment]
@HighCommander4 The contents of compile_commands.json in the same folder as these files is:

```json
[
    {
        "arguments": ["-std=c17"],
    },
]

```

**Comment by sam-mccall** [comment]
> it now does add missing header for printf, but it in the process overwrites the function prototype, and duplicates the printf statement.

This sounds like a race where the fix is being applied to the wrong version of the file or so. I think we do have such cases, but it's not really related to this bug.https://reviews.llvm.org/D115232

> As a hack for now, can I create an empty file and add all the C standard library header files for indexer to recognize it?

This should work. If you're building locally you could also try out https://reviews.llvm.org/D115232 which automatically indexes the standard library.

**Comment by HighCommander4** [comment]
> The contents of compile_commands.json in the same folder as these files is:
> 
> ```json
> [
>     {
>         "arguments": ["-std=c17"],
>     },
> ]
> ```

I don't think that's well-formed (see [spec](https://clang.llvm.org/docs/JSONCompilationDatabase.html)). Every entry must have `directory` and `file` keys, as well as either `command` or `argument`. And there needs to be an entry for every translation unit that you want included in the index.

Note, this file is typically not hand-written, but generated. See the [Project setup](https://clangd.llvm.org/installation#project-setup) of the clangd docs for some ways how.

> As a hack for now, can I create an empty file and add all the C standard library header files for indexer to recognize it? 

Yep. (Be sure it has an entry in the `compile_commands.json`.) This is what we plan to do automatically (without actually creating a file in the filesystem) in #403.

**Comment by reportaman** [comment]
@sam-mccall Yep I built clangd locally, please excuse me for not knowing this, but its not clear if and how to apply that patch. Is it a git branch I need to checkout and then build clangd again? (I know the build from source commands from clangd page: https://github.com/llvm/llvm-project/tree/main/clang-tools-extra/clangd).

**Comment by reportaman** [comment]
Ah the patch is downloadable, never mind. ``git apply --check x.patch.diff``, ``git apply --stat x.patch.diff``, and finally ``git apply  x.patch.diff``. One last question before I try that:  What's the exact syntax of setting ``Index.StandardLibrary`` in compile_commands.json?

**Comment by sam-mccall** [comment]
You can click "download raw diff" and run `patch -p1 < D115232.diff` in the `llvm-project` directory.
This will modify your local sources, you can commit it to a branch or so if you like.

**Comment by sam-mccall** [comment]
What's the exact syntax of setting Index.StandardLibrary in compile_commands.json?

It goes in the config file, not in compile_commands.json:
```
$ cat ~/.config/clangd/config.yaml
Index:
  StandardLibrary: Yes
```

**Comment by reportaman** [comment]
@sam-mccall Same saga for time.h and assert.h after rebuilding clangd with the patch, reinstalling coc-clangd, and setting that config file. There is no fix for assert(), and time().

```c
#include <stdio.h>
#include <stdlib.h>
int main(void){
    assert(0);
    time(NULL);
    printf("\n");
    return EXIT_SUCCESS;
}

```

**Comment by sam-mccall** [comment]
OK, there's more going on here...
(The bug description says "like printf" - printf is a special case! I didn't realize this was happening for functions in general).

- the warning "implicit declaration of function..." (`ext_implicit_function_decl` and friends) is not on our list to insert includes for. This is easy to add.
- we have a check to only insert includes for errors, and this one is a warning. I'm not sure why we have this check, we can remove it.
- we use clang's typo-correction hook to observe unresolved identifiers, but clang avoids running typo correction in this case unless the diagnostic is an error. We'll want to override this somehow, I think this is the main open question.
- clang runs typo correction *after* producing this diagnostic, and publishes the results as a note. I think we can just reorder this, it makes the clang code a little awkward is all.

If I hack in fixes to all four of these, the fixes start working again.

**Comment by sam-mccall** [comment]
> Nor is there a fix for EXIT_SUCCESS (if I did not have stdlib.h included while opening file to begin with).

There are probably issues with macro indexing, (there usually are). I'll have to come back to that one later.

**Comment by sam-mccall** [comment]
https://reviews.llvm.org/D115490 for C functions.

**Comment by reportaman** [comment]
@sam-mccall Even with the patch, I don't see a code action/ fix in coc-clangd for assert() or time()...

**Comment by sam-mccall** [comment]
@reportaman Both assert() and time() work for me with a slight caveat: if this is the first file you open, you won't see the code action until after you modify some code. This is a sequencing issue: the code actions are produced along with diagnostics based on the index available at the time. For the very first diagnostic pass of the very first file you open, the index probably isn't available yet.

My setup is both patches applied, and standard library indexing enabled in config.yaml.

**Comment by sam-mccall** [comment]
Sorry for the delay and thanks for being persistent.
The C-specific part of this seems to be fixed - I'm now seeing the same behavior between C and C++.

Automatic standard library indexing hasn't landed (yet).
